File: sast-java/src/main/java/com/sast/astbenchmark/case_language_maturity/accuracy/context_sensitive/argument_return_value_passing/Chained_Call_001_T.java
package com.sast.astbenchmark.case_language_maturity.accuracy.context_sensitive.argument_return_value_passing;

import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

// evaluation information start
// real case = true
// evaluation item = 准确度->上下文敏感分析->参数/返回值传递
// scene introduction = 链式调用
// level = 2
// bind_url = accuracy/context_sensitive/argument_return_value_passing/Chained_Call_001_T
// date = 2025-12-10 15:15:00
// evaluation information end

@RestController()
@RequestMapping("accuracy/context_sensitive/argument_return_value_passing")
public class Chained_Call_001_T {
  @GetMapping("Chained_Call_001_T/{cmd}")
  public Map<String, Object> chained_call_001_t(@PathVariable String cmd) {
    Map<String, Object> modelMap = new HashMap<>();
    try {
      // 场景特点：链式调用中的返回值传递，污点通过链式调用传递
      String result = new CommandProcessor().setCommand(cmd).process();
      Runtime.getRuntime().exec(result);
      modelMap.put("status", "success");
    } catch (Exception e) {
      modelMap.put("status", "error");
    }
    return modelMap;
  }

  static class CommandProcessor {
    private String command;

    public CommandProcessor setCommand(String command) {
      this.command = command;
      return this;
    }

    public String process() {
      return command;
    }
  }
}

File: sast-php/case/completeness/single_app_tracing/function_call/chained_call_001_T.php
<?php
// evaluation information start
// real case = true
// evaluation item = 完整度->单应用追踪->函数调用
// scene introduction = 链式方法调用传递污点
// level = 2
// bind_url = completeness/single_app_tracing/function_call/chained_call_001_T
// evaluation information end

class Builder {
    private $data = "";

    public function setData($val) {
        $this->data = $val;
        return $this;
    }

    public function build() {
        return $this->data;
    }
}

function chained_call_001_T($__taint_src) {
    $builder = new Builder();
    $result = $builder->setData($__taint_src)->build();
    __taint_sink($result);
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
chained_call_001_T($taint_src);
